internal/pkg/scaffold/olm-catalog/descriptor.go

```diff
@@ -0,0 +1,544 @@
+		if pkg.Path != apiPkg && !strings.HasPrefix(pkg.Path, "./") {
+			continue
+		}
+		for _, t := range pkg.Types {
+			pkgTypes = append(pkgTypes, t)
+			if t.Name.Name == kind {
+				for _, m := range t.Members {
+					path := parsePathFromJSONTags(m.Tags)
+					if path == typeSpec {
+						spec = m.Type
+					} else if path == typeStatus {
+						status = m.Type
+					}
+					if spec != nil && status != nil {
+						break
+					}
+				}
+			}
+		}
+	}
+	if spec == nil {
+		return nil, nil, nil, errors.Errorf("no spec found in type %s", kind)
+	}
+	if status == nil {
+		return nil, nil, nil, errors.Errorf("no status found in type %s", kind)
+	}
+	if len(pkgTypes) == 0 {
+		return nil, nil, nil, errors.Errorf("no package types found in API %s", apiPkg)
+	}
+	return spec, status, pkgTypes, nil
+}
+
+type descriptorType = string
+
+const (
+	typeSpec   descriptorType = "spec"
+	typeStatus descriptorType = "status"
+)
+
+type descriptor struct {
+	include     bool
+	parentType  *types.Type
+	member      types.Member
+	descType    descriptorType
+	description string
+	displayName string
+	path        string
+	xdesc       []string
+}
+
+func sortDescriptors(ds []descriptor) []descriptor {
+	sort.Slice(ds, func(i, j int) bool {
+		return ds[i].path < ds[j].path
+	})
+	return ds
+}
+
+type parsedCRDDescriptions struct {
+	descriptors []descriptor
+	displayName string
+	resources   []olmapiv1alpha1.APIResourceReference
+}
+
+func sortResources(rs []olmapiv1alpha1.APIResourceReference) []olmapiv1alpha1.APIResourceReference {
+	sort.Slice(rs, func(i, j int) bool {
+		return rs[i].Kind < rs[j].Kind
+	})
+	return rs
+}
+
+func wrapParseErr(err error) error {
+	return errors.Wrap(err, "error parsing csv-gen annotation")
+}
+
+// parseCSVGenAnnotations parses all descriptor annotations from comments,
+// each of which should contain one spec.customresourcedefinitions.owned entry.
+// field Once all comments have been parsed, the entry is added to a
+// parsedCRDDescriptions.
+func parseCSVGenAnnotations(comments []string) (pd parsedCRDDescriptions, err error) {
+	tags := types.ExtractCommentTags(csvgenPrefix, comments)
+	specd, statusd := descriptor{descType: typeSpec}, descriptor{descType: typeStatus}
+	for path, vals := range tags {
+		pathElems, err := annotations.SplitPath(path)
+		if err != nil {
+			return pd, wrapParseErr(err)
+		}
+		parentPathElem, childPathElems := pathElems[0], pathElems[1:]
+		switch parentPathElem {
+		case "customresourcedefinitions":
+			switch childPathElems[0] {
+			case "specDescriptors":
+				err = parseDescriptor(&specd, childPathElems, vals[0])
+				if err != nil {
+					return pd, wrapParseErr(err)
+				}
+			case "statusDescriptors":
+				err = parseDescriptor(&statusd, childPathElems, vals[0])
+				if err != nil {
+					return pd, wrapParseErr(err)
+				}
+			case "displayName":
+				pd.displayName, err = strconv.Unquote(vals[0])
+				if err != nil {
+					return pd, errors.Wrapf(err, "error unquoting %s", vals[0])
```

Comment:
This is supposed to be wrapped with `wrapParseErr()` as well right? Since that's common to this function.

Another way you can wrap a common prefix to all errors of a function is by doing it once at the top in a defer function:
```Go
defer func() {
    if err != nil {
        err = parseResource(err)
    }
}()
```

And then just keeping this as `return pd, errors.Wrapf(err, "error unquoting %s", vals[0])`